barretenberg/cpp/src/barretenberg/plonk/proof_system/widgets/transition_widgets/elliptic_widget.hpp

```diff
@@ -106,46 +107,39 @@ template <class Field, class Getters, typename PolyContainer> class EllipticKern
         const Field& x_3 = Getters::template get_value<EvaluationType::SHIFTED, PolynomialIndex::W_2>(polynomials, i);
         const Field& y_3 = Getters::template get_value<EvaluationType::SHIFTED, PolynomialIndex::W_3>(polynomials, i);
 
-        // Endomorphism coefficient for when we add and multiply by beta at the same time
-        const Field& q_beta =
-            Getters::template get_value<EvaluationType::NON_SHIFTED, PolynomialIndex::Q_3>(polynomials, i);
-        // Square of endomorphism coefficient
-        const Field& q_beta_sqr =
-            Getters::template get_value<EvaluationType::NON_SHIFTED, PolynomialIndex::Q_4>(polynomials, i);
         // sign
         const Field& q_sign =
             Getters::template get_value<EvaluationType::NON_SHIFTED, PolynomialIndex::Q_1>(polynomials, i);
 
-        // TODO: Can this be implemented more efficiently?
-        // It seems that Zac wanted to group the elements by selectors to use several linear terms initially,
-        // but in the end we are using one, so there is no reason why we can't optimize computation in another way
-
-        Field beta_term = -x_2 * x_1 * (x_3 + x_3 + x_1); // -x_1 * x_2 * (2 * x_3 + x_1)
-        Field beta_sqr_term = x_2.sqr();                  // x_2^2
-        Field leftovers = beta_sqr_term;                  // x_2^2
-        beta_sqr_term *= (x_3 - x_1);                     // x_2^2 * (x_3 - x_1)
-        Field sign_term = y_2 * y_1;                      // y_1 * y_2
-        sign_term += sign_term;                           // 2 * y_1 * y_2
-        beta_term *= q_beta;                              // -β * x_1 * x_2 * (2 * x_3 + x_1)
-        beta_sqr_term *= q_beta_sqr;                      // β^2 * x_2^2 * (x_3 - x_1)
-        sign_term *= q_sign;                              // 2 * y_1 * y_2 * sign
-        leftovers *= x_2;                                 // x_2^3
-        leftovers += x_1.sqr() * (x_3 + x_1);             // x_2^3 + x_1 * (x_3 + x_1)
-        leftovers -= (y_2.sqr() + y_1.sqr());             // x_2^3 + x_1 * (x_3 + x_1) - y_2^2 - y_1^2
-
-        // Can be found in class description
-        Field x_identity = beta_term + beta_sqr_term + sign_term + leftovers;
-        x_identity *= challenges.alpha_powers[0];
-
-        beta_term = x_2 * (y_3 + y_1) * q_beta;  // β * x_2 * (y_3 + y_1)
-        sign_term = -y_2 * (x_1 - x_3) * q_sign; // - signt * y_2 * (x_1 - x_3)
-        // TODO: remove extra additions if we decide to stay with this implementation
-        leftovers = -x_1 * (y_3 + y_1) + y_1 * (x_1 - x_3); // -x_1 * y_3 - x_1 * y_1 + y_1 * x_1 - y_1 * x_3
-
-        Field y_identity = beta_term + sign_term + leftovers;
-        y_identity *= challenges.alpha_powers[1];
-
-        linear_terms[0] = x_identity + y_identity;
+        const Field& q_is_double =
+            Getters::template get_value<EvaluationType::NON_SHIFTED, PolynomialIndex::Q_M>(polynomials, i);
```

Comment:
This assumes that q_elliptic  = 1, so whether its a doubling gate depends on whether q_m = 1.

It computes both the doubling and addition relation and then turns either one off depending on whether q_m = 1

Reply:
Going to come back and check the elliptic equations; noting that we can reuse the challenges `challenges.alpha_powers[0]` and `challenges.alpha_powers[1]` between the two relations because one of the realtions will be zeroed out

Reply:
just modified the code to use be more efficient and reuse challenges